tasks/eva_2779/correct_assembly_error_in_rat.py

```diff
@@ -0,0 +1,268 @@
+    cursor = collection.with_options(read_concern=ReadConcern("majority")) \
+                       .find(filter_criteria, no_cursor_timeout=True)
+    records = []
+    try:
+        for result in cursor:
+            records.append(result)
+            if len(records) == batch_size:
+                yield records
+                records = []
+        yield records
+    except Exception as e:
+        logger.exception(traceback.format_exc())
+        raise e
+    finally:
+        cursor.close()
+
+
+def replace_document_with_correct_information(mongo_source, collection_name, id_creation_func, filter_criteria,
+                                              correction_map, batch_size):
+    collection = mongo_source.mongo_handle[mongo_source.db_name][collection_name]
+
+    total_inserted, total_dropped = 0, 0
+    try:
+        for batch_of_variant in find_documents_in_batch(mongo_source, collection_name, filter_criteria, batch_size):
+            insert_statements = []
+            drop_statements = []
+            for variant in batch_of_variant:
+                original_id = id_creation_func(variant)
+                assert variant['_id'] == original_id, "Original id is different from the one calculated %s != %s" % (
+                    variant['_id'], original_id)
+                for key in correction_map:
+                    variant[key] = correction_map[key]
+                variant['_id'] = id_creation_func(variant)
+                insert_statements.append(pymongo.InsertOne(variant))
+                drop_statements.append(pymongo.DeleteOne({'_id': original_id}))
+            result_insert = collection.with_options(write_concern=WriteConcern(w="majority", wtimeout=1200000))\
+                                      .bulk_write(requests=insert_statements, ordered=False)
+            total_inserted += result_insert.inserted_count
+            result_drop = collection.with_options(write_concern=WriteConcern(w="majority", wtimeout=1200000)) \
+                                    .bulk_write(requests=drop_statements, ordered=False)
+            total_dropped += result_drop.deleted_count
+        logger.info(f'{total_inserted} new documents inserted in {collection_name}')
+        logger.info(f'{total_dropped} old documents dropped in {collection_name}')
+    except Exception as e:
+        print(traceback.format_exc())
+        raise e
+    return total_inserted
+
+
+def update_operation_entities(mongo_source, collection_name, id_creation_func, filter_criteria, correction_map, batch_size):
+    collection = mongo_source.mongo_handle[mongo_source.db_name][collection_name]
+    total_updated = 0
+    try:
+        for batch_of_variant in find_documents_in_batch(mongo_source, collection_name, filter_criteria, batch_size):
+            update_statements = []
+            for variant in batch_of_variant:
+                filter_dict = {'_id': variant['_id']}
+                filter_dict.update(filter_criteria)
```

Comment:
I am wondering why we need an additional criteria when we already have the ID...

Reply:
We don't anymore. In a previous version of the script I was using [the $ notation](https://www.mongodb.com/docs/v4.0/reference/operator/update/positional/index.html?_ga=2.68455185.1468278035.1648464773-883181489.1633612269) to update the sub arrays which required that the field replaced was present in the filter. But it turned out to be too complicated.